src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/SpanNDHelpers.cs

```diff
@@ -0,0 +1,508 @@
+                }
+                else if (sizeof(T) == sizeof(long))
+                {
+                    Reverse(ref Unsafe.As<T, long>(ref elements), length);
+                    return;
+                }
+            }
+
+            ReverseInner(ref elements, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void ReverseInner<T>(ref T elements, nuint length)
+        {
+            Debug.Assert(length > 1);
+
+            ref T first = ref elements;
+            ref T last = ref Unsafe.Subtract(ref Unsafe.Add(ref first, length), 1);
+            do
+            {
+                T temp = first;
+                first = last;
+                last = temp;
+                first = ref Unsafe.Add(ref first, 1);
+                last = ref Unsafe.Subtract(ref last, 1);
+            } while (Unsafe.IsAddressLessThan(ref first, ref last));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static nint CalculateTotalLength(ref nint[] lengths)
+        {
+            nint totalLength = 1;
+            for (int i = 0; i < lengths.Length; i++)
+            {
+                totalLength *= lengths[i];
+            }
+
+            return totalLength;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static nint CalculateTotalLength(ReadOnlySpan<nint> lengths)
+        {
+            nint totalLength = 1;
+            for (int i = 0; i < lengths.Length; i++)
+            {
+                totalLength *= lengths[i];
+            }
+
+            return totalLength;
+        }
+
+        /// <summary>
+        /// Gets the set of strides that can be used to calculate the offset of n-dimensions in a 1-dimensional layout
+        /// </summary>
+        /// <returns></returns>
+        public static nint[] CalculateStrides(int rank, ReadOnlySpan<nint> lengths)
+        {
+            nint[] strides = new nint[rank];
+
+            nint stride = 1;
+
+            for (int i = strides.Length - 1; i >= 0; i--)
+            {
+                strides[i] = stride;
+                stride *= lengths[i];
+            }
+
+            return strides;
+        }
+
+        /// <summary>
+        /// Calculates the 1-d index for n-d indices in layout specified by strides.
+        /// </summary>
+        /// <param name="indices"></param>
+        /// <param name="strides"></param>
+        /// <param name="lengths"></param>
+        /// <returns></returns>
+        public static nint GetIndex(ReadOnlySpan<nint> indices, ReadOnlySpan<nint> strides, ReadOnlySpan<nint> lengths)
+        {
+            Debug.Assert(strides.Length == indices.Length);
+
+            nint index = 0;
+            for (int i = 0; i < indices.Length; i++)
+            {
+                if (indices[i] >= lengths[i])
+                    ThrowHelper.ThrowIndexOutOfRangeException();
+                index += strides[i] * (indices[i]);
```

Comment:
We're in general missing argument validation or assert checks for many of these APIs.

Several of them are directly taking user input and will do the wrong thing for things like a negative index.

Reply:
So at least all of the helpers in here are internal only and shouldn't be taking anything directly from the user. I'll make sure though that anywhere we call these the input has been validated first.